### ISSUE TITLE
v0.5.3 ArgumentError: range must be non-empty

### ISSUE THREAD

**Issue by fkrauer** [itsdfish/DifferentialEvolutionMCMC.jl#37]
Hi @itsdfish 

I have just updated to v0.5.3 and tried to run your example (https://github.com/itsdfish/DifferentialEvolutionMCMC.jl/blob/master/Examples/Multivariate_Guassian_Example.jl) for the snooker update, and it throws an error:

```
ERROR: LoadError: UndefVarError: resample not defined
```

When I omit the sample=resample argument, it throws another error:
```
got unsupported keyword arguments "n_initial", "θsnooker"
```

Thanks

**Comment by itsdfish** [comment]
Hello,

Hmmm. That is strange. The example works on my computer with v0.5.3. Note that resample is [exported](https://github.com/itsdfish/DifferentialEvolutionMCMC.jl/blob/88e82529613d47fc7527d813c7fd0eae5dbf8704/src/DifferentialEvolutionMCMC.jl#L10) to the main module. So it should be available. I suspect that in the process of updating your packages, another package may have downgraded  DifferentialEvolutionMCMC to an earlier version. Can you confirm that you are still using v0.5.3?

**Comment by fkrauer** [comment]
Yes I have v.0.5.3, but it seems to work now after opening a new REPL. Maybe some pre-compiling problem.

**Comment by itsdfish** [comment]
I'm glad it was an easy fix!

**Comment by fkrauer** [comment]
Actually, my code throws another error now since the update to 0.5.3 with the snooker updating. It occurs both in single and multi-threading mode. I have pasted the code and error below. I couldn't make much sense of it, but it may be related to some operation including rand()? TIA! 

error:
```
ERROR: LoadError: TaskFailedException:
ArgumentError: range must be non-empty
Stacktrace:
 [1] SamplerRangeNDL at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\generation.jl:315 [inlined] [2] Sampler at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\generation.jl:183 [inlined]        
 [3] Sampler at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\Random.jl:139 [inlined]
 [4] rand at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\Random.jl:253 [inlined]
 [5] rand at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Random\src\Random.jl:258 [inlined]
 [6] resample(::DE{NTuple{4,Tuple{Float64,Float64}},typeof(random_gamma),typeof(DifferentialEvolutionMCMC.Metropolis_Hastings_update!),typeof(compute_posterior!),typeof(resample)}, ::Array{Particle{Float64},1}, ::Int64, ::Bool) at C:\Users\micky\.julia\packages\DifferentialEvolutionMCMC\gOwqH\src\crossover.jl:33
 [7] random_gamma(::DE{NTuple{4,Tuple{Float64,Float64}},typeof(random_gamma),typeof(DifferentialEvolutionMCMC.Metropolis_Hastings_update!),typeof(compute_posterior!),typeof(resample)}, ::Particle{Float64}, ::Array{Particle{Float64},1}) at C:\Users\micky\.julia\packages\DifferentialEvolutionMCMC\gOwqH\src\crossover.jl:57
 [8] crossover!(::DEModel{DifferentialEvolutionMCMC.var"#3#4"{typeof(loglik),Array{Float64,1},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},Tuple{ODEProblem{LArray{Float64,1,Array{Float64,1},(:S, :I, :R, :C)},Tuple{Float64,Float64},true,NamedTuple{(:β, :η, :φ, :ω, :μ, :σ),Tuple{Float64,Float64,Int64,Float64,Float64,Float64}},ODEFunction{true,typeof(SIR!),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,typeof(SciMLBase.DEFAULT_OBSERVED),Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},SciMLBase.StandardODEProblem},NamedTuple{(:μ, :σ),Tuple{Float64,Float64}},NTuple{4,Symbol},NamedTuple{(:abstol, :reltol, :maxiters, :saveat, :solver),Tuple{Float64,Float64,Float64,Float64,CompositeAlgorithm{Tuple{Tsit5,Rosenbrock23{0,true,DefaultLinSolve,DataType}},AutoSwitch{Tsit5,Rosenbrock23{0,true,DefaultLinSolve,DataType},Rational{Int64},Int64}}}}}},NTuple{4,Tuple{Uniform{Float64}}},NTuple{4,String}}, ::DE{NTuple{4,Tuple{Float64,Float64}},typeof(random_gamma),typeof(DifferentialEvolutionMCMC.Metropolis_Hastings_update!),typeof(compute_posterior!),typeof(resample)}, ::Array{Particle{Float64},1}) at C:\Users\micky\.julia\packages\DifferentialEvolutionMCMC\gOwqH\src\crossover.jl:18
 [9] mutate_or_crossover!(::DEModel{DifferentialEvolutionMCMC.var"#3#4"{typeof(loglik),Array{Float64,1},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},Tuple{ODEProblem{LArray{Float64,1,Array{Float64,1},(:S, :I, :R, :C)},Tuple{Float64,Float64},true,NamedTuple{(:β, :η, :φ, :ω, :μ, :σ),Tuple{Float64,Float64,Int64,Float64,Float64,Float64}},ODEFunction{true,typeof(SIR!),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,typeof(SciMLBase.DEFAULT_OBSERVED),Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},SciMLBase.StandardODEProblem},NamedTuple{(:μ, :σ),Tuple{Float64,Float64}},NTuple{4,Symbol},NamedTuple{(:abstol, :reltol, :maxiters, :saveat, :solver),Tuple{Float64,Float64,Float64,Float64,CompositeAlgorithm{Tuple{Tsit5,Rosenbrock23{0,true,DefaultLinSolve,DataType}},AutoSwitch{Tsit5,Rosenbrock23{0,true,DefaultLinSolve,DataType},Rational{Int64},Int64}}}}}},NTuple{4,Tuple{Uniform{Float64}}},NTuple{4,String}}, ::DE{NTuple{4,Tuple{Float64,Float64}},typeof(random_gamma),typeof(DifferentialEvolutionMCMC.Metropolis_Hastings_update!),typeof(compute_posterior!),typeof(resample)}, ::Array{Particle{Float64},1}, ::UInt64) at C:\Users\micky\.julia\packages\DifferentialEvolutionMCMC\gOwqH\src\main.jl:103
 [10] macro expansion at C:\Users\micky\.julia\packages\DifferentialEvolutionMCMC\gOwqH\src\main.jl:89 [inlined]
 [11] (::DifferentialEvolutionMCMC.var"#54#threadsfor_fun#11"{DEModel{DifferentialEvolutionMCMC.var"#3#4"{typeof(loglik),Array{Float64,1},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},Tuple{ODEProblem{LArray{Float64,1,Array{Float64,1},(:S, :I, :R, :C)},Tuple{Float64,Float64},true,NamedTuple{(:β, :η, :φ, :ω, :μ, :σ),Tuple{Float64,Float64,Int64,Float64,Float64,Float64}},ODEFunction{true,typeof(SIR!),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,typeof(SciMLBase.DEFAULT_OBSERVED),Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},SciMLBase.StandardODEProblem},NamedTuple{(:μ, :σ),Tuple{Float64,Float64}},NTuple{4,Symbol},NamedTuple{(:abstol, :reltol, :maxiters, :saveat, :solver),Tuple{Float64,Float64,Float64,Float64,CompositeAlgorithm{Tuple{Tsit5,Rosenbrock23{0,true,DefaultLinSolve,DataType}},AutoSwitch{Tsit5,Rosenbrock23{0,true,DefaultLinSolve,DataType},Rational{Int64},Int64}}}}}},NTuple{4,Tuple{Uniform{Float64}}},NTuple{4,String}},DE{NTuple{4,Tuple{Float64,Float64}},typeof(random_gamma),typeof(DifferentialEvolutionMCMC.Metropolis_Hastings_update!),typeof(compute_posterior!),typeof(resample)},Array{Array{Particle{Float64},1},1},Array{UInt64,1},UnitRange{Int64}})(::Bool) at .\threadingconstructs.jl:81
 [12] (::DifferentialEvolutionMCMC.var"#54#threadsfor_fun#11"{DEModel{DifferentialEvolutionMCMC.var"#3#4"{typeof(loglik),Array{Float64,1},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},Tuple{ODEProblem{LArray{Float64,1,Array{Float64,1},(:S, :I, :R, :C)},Tuple{Float64,Float64},true,NamedTuple{(:β, :η, :φ, :ω, :μ, :σ),Tuple{Float64,Float64,Int64,Float64,Float64,Float64}},ODEFunction{true,typeof(SIR!),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,typeof(SciMLBase.DEFAULT_OBSERVED),Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},SciMLBase.StandardODEProblem},NamedTuple{(:μ, :σ),Tuple{Float64,Float64}},NTuple{4,Symbol},NamedTuple{(:abstol, :reltol, :maxiters, :saveat, :solver),Tuple{Float64,Float64,Float64,Float64,CompositeAlgorithm{Tuple{Tsit5,Rosenbrock23{0,true,DefaultLinSolve,DataType}},AutoSwitch{Tsit5,Rosenbrock23{0,true,DefaultLinSolve,DataType},Rational{Int64},Int64}}}}}},NTuple{4,Tuple{Uniform{Float64}}},NTuple{4,String}},DE{NTuple{4,Tuple{Float64,Float64}},typeof(random_gamma),typeof(DifferentialEvolutionMCMC.Metropolis_Hastings_update!),typeof(compute_posterior!),typeof(resample)},Array{Array{Particle{Float64},1},1},Array{UInt64,1},UnitRange{Int64}})() at .\threadingconstructs.jl:48
```

code:
```
using DifferentialEquations, StatsBase, StatsPlots, DifferentialEvolutionMCMC, Distributions, Random, StatsBase, LabelledArrays

Random.seed!(42)

# ODE model: simple SIR model with seasonally forced contact rate
function SIR!(du,u,p,t)

    # states
    S, I, R = u
    N = S + I + R
  
    # params
    β = p.β
    η = p.η
    φ = p.φ
    ω = p.ω
    μ = p.μ
    σ = p.σ
      
    # FOI
    βeff = β * (1.0+η*sin(2.0*π*(t-φ)/365.0))
    λ = βeff*I/N

    # change in states
    du[1] = (μ*N - λ*S - μ*S + ω*R)
    du[2] = (λ*S - σ*I - μ*I)
    du[3] = (σ*I - μ*R - ω*R)
    du[4] = (σ*I) # cumulative incidence
  
end
  
# Calculate incident cases from cumulative incidence
function get_incidence(sol)
    incidence = [sol.u[t].C - sol.u[t-1].C  for t = 2:length(sol.u)]
    incidence = ifelse.(incidence .< 0.0, 0.0, incidence)
    return incidence
end

# Solver settings
tmin = 0.0
tmax = 10.0*365.0
tspan = (tmin, tmax)
solvsettings = (abstol = 1.0e-6, reltol = 1.0e-6, 
                maxiters = 1e5, saveat = 7.0,
                solver = AutoTsit5(Rosenbrock23())) #AutoVern7(Rodas5()))


# Initiate ODE problem
theta_fix = (μ=1.0/(80*365),σ=1.0/5.0)
theta_est = (β=0.28,η=0.07,φ=20,ω=1.0/365)
p = merge(theta_est, theta_fix)
u0 = @LArray [9998.0,1.0,1.0,1.0] (:S,:I,:R,:C)

# Initiate ODE problem and run solution with dummy vars
problem = ODEProblem(SIR!,u0,tspan,p)
sol = solve(problem, 
            solvsettings.solver, 
            abstol=solvsettings.abstol, 
            reltol=solvsettings.reltol, 
            maxiters=solvsettings.maxiters,
            isoutofdomain=(u,p,t)->any(x->x<0.0,u), 
            saveat=solvsettings.saveat)
#plot(sol)
incidence = get_incidence(sol)
#plot(incidence)


#soldf = DataFrame(sol)
#N = sum(solarray,2) 
#S = exp.(soldf.value1)
#I = exp.(soldf.value2)
#R = exp.(soldf.value3)
#N = S .+ I .+ R

# Fake some data from model
data = round.(incidence .* rand(Uniform(0.5,1.5), size(incidence)))
#plot(incidence, legend = false); scatter!(data,legend = false)
#data = @. round(incidence * rand(Uniform(0.5,1.5), size(incidence)))

# Fit model to fake data

# Priors
priors = (
    β = (Uniform(1e-3,1.0),), 
    η = (Uniform(0.0,1.0),),
    ω = (Uniform(1.0/(5.0*365.0),1),),
    φ = (Uniform(0.0,364.0),)
)

# Upper and lower bounds (in this case equivalent to arguments of the Uniform)
bounds = (
    (1e-3,1.0),
    (0.0,1.0),
    (1.0/(5.0*365.0),1.0),
    (0.0,364.0)
)

parnames = keys(priors)

# Log Likelihood function
#function loglik(data, problem, theta_fix, parnames, θ...)

function loglik(data, problem, theta_fix, parnames, solvsettings, θ...)

    theta_est = NamedTuple{parnames}(θ)
    p_new = merge(theta_fix, theta_est)

    # Update problem and solve ODEs
    problem_new = remake(problem, p=p_new)
    sol_new = solve(problem_new, 
                    solvsettings.solver, 
                    abstol=solvsettings.abstol, 
                    reltol=solvsettings.reltol, 
                    maxiters=solvsettings.maxiters,
                    isoutofdomain=(u,p,t)->any(x->x<0.0,u), 
                    saveat=solvsettings.saveat)

    # due to numerical instability of the diff eqs, some states can become negative. In this case, return -Inf
    #if minimum(sol_new) < 0.0
    #   return -Inf
    #else

    # calculate incidence
    incidence = get_incidence(sol_new)

    ll = sum(logpdf.(Poisson.(incidence), data))
    return ll
end

# test function
loglik(data, problem, theta_fix, parnames, solvsettings, theta_est...)

# Run DE MCMC
burnin=500
n_iter = 1000
n_groups = 3 # default is 4
Np = 3 # N of particles per group.Default is nparam*3
model = DEModel(problem, theta_fix, parnames, solvsettings; priors, model=loglik, data=data)
de = DE(; priors=priors, n_groups=n_groups, Np=Np, sample=resample, θsnooker=.1, burnin=burnin, α=0.1, β=0.1, ϵ=0.001, σ=0.05, κ=1.0, 
bounds=bounds)
@time chains = sample(model, de, MCMCThreads(), n_iter, progress=true)

```

**Comment by itsdfish** [comment]
Ok. I see the problem. Your code is missing a value for the keyword argument `n_initial`, which is the initial number of samples from the prior for the resampling option. A common value is 10 times the number of parameters.  Another issue is that snooker updating requires Np > 3. There might be a way to reduce it to three. 

Also, for when variable names match the keyword argument, you do not have to write `keyword = keyword`. It may shorten a few statements if you would like. 

I will keep this open for now because I forgot to document `n_initial`. 

```
# Run DE MCMC
burnin=500
n_iter = 1000
n_groups = 3 # default is 4
Np = 4 # N of particles per group.Default is nparam*3
model = DEModel(problem, theta_fix, parnames, solvsettings; priors, model=loglik, data)
de = DE(; priors, n_groups, Np, sample=resample, θsnooker=.1, burnin, 
    α=0.1, β=0.1, ϵ=0.001, σ=0.05, κ=1.0, bounds, n_initial=10*length(theta_est))
@time chains = sample(model, de, MCMCThreads(), n_iter, progress=true)
```

**Comment by fkrauer** [comment]
ah yes, it works now, thanks. 

Regarding Np, I thought that for the DEzs algorithm with the snooker update, the number of chains per group can be reduced to 3, irrespective of the number of parameters (see Ter Braak, C.J.F., Vrugt, J.A., 2008. Differential Evolution Markov Chain with snooker updater and fewer chains. Statistics and Computing 18, 435–446.. doi:10.1007/s11222-008-9104-9), but I may be wrong. 

Also, is the thinning rate automatically set to 10 when using the snooker update?

**Comment by itsdfish** [comment]
Yes. You are right. The paper states that a minimum of three particles is required. It was not clear to me which particles can be sampled with replacement. Upon further thought, I think I know how to reduce it to 3. I will look into it shortly. The default for thinning is 1, but its not implemented. I will add that too.

**Comment by itsdfish** [comment]
@fkrauer, I updated the documentation and changed `resample` so that it requires only 3 chains. These updates are reflected in v0.5.4. 

I decided against adding a feature for thinning for the time being. Based on a few papers that I read, thinning does not typically improve the quality of the MCMC samples. It can minimize storage requirements for very large models. However, the resampling method mostly eliminates this problem.
